### ISSUE TITLE
Ask a question?

### ISSUE THREAD

**Issue by Njzjhd** [ZongSingHuang/Binary-Whale-Optimization-Algorithm#1]
def Breastcancer_test(x):
    if x.ndim==1:
        x = x.reshape(1, -1) 
    loss = np.zeros(x.shape[0]) #array([0.])
    
    for i in range(x.shape[0]):
        if np.sum(x[i, :])>0:
            knn = KNeighborsClassifier(n_neighbors=5).fit(X_train[:, x[i, :]], y_train)
            score = accuracy_score(knn.predict(X_test[:, x[i, :]]), y_test)
            loss[i] = 0.99*(1-score) + 0.01*(np.sum(x[i, :])/X_train.shape[1])
        else:
            loss[i] = np.inf
            print(666)
    return loss

For the fitness function, is there a computational error in the for loop, there should be only one value, right? Is it possible to introduce cross-validation to further remedy the defect?

**Comment by ZongSingHuang** [comment]
Thanks for your response

1. if np.all(x[i])==0, meaning no feature selected, so I put np.inf to loss[i]
2. I'm following the literature experiment setting, so main_5050.py have not adopted kfold
3. Finally, main_5050.py have adopted kfold

![image](https://user-images.githubusercontent.com/58209813/105466915-e848c180-5ccf-11eb-888f-9c01cb12c7a5.png)

**Comment by Njzjhd** [comment]
Thank you very mach!

**Comment by Njzjhd** [comment]
X_train[:, optimizer.gBest_X]
This code does not capture valid features and may have problems. Usually 1 is selected, but this is not the case.

**Comment by Njzjhd** [comment]
I have transformed the above problem, which is a bit complicated, but can effectively solve the problem

#Get the index equal to 1
def unique_index(L,f):
        return [i for (i,v) in enumerate(L) if v==f]
t_cols = unique_index(tt,1)

#Add features with index equal to 1 to a new dataset
tr_cols = X_train.columns.to_list()
tr_add = []
for i in range(len(t_cols)):
         tr_add.append(tr_cols[t_cols[i]])

**Comment by ZongSingHuang** [comment]
I have fix this problem, thank you
![image](https://user-images.githubusercontent.com/58209813/105492607-26a3a800-5cf3-11eb-877a-93daf875f8cb.png)
![image](https://user-images.githubusercontent.com/58209813/105492927-9fa2ff80-5cf3-11eb-98a9-c5a1494e16e4.png)

**Comment by ZongSingHuang** [comment]
I have fix this problem, thank you
![image](https://user-images.githubusercontent.com/58209813/105492607-26a3a800-5cf3-11eb-877a-93daf875f8cb.png)

**Comment by Njzjhd** [comment]
For " for i in range(x.shape[0]):", i seems to be equal to 0 only. It seems to be impossible to implement a for loop, the author can do the test. Also what is the meaning of the code " if x.ndim==1: x = x.reshape(1, -1) loss = np.zeros(x.shape[0]) #array([0.])"? This code must be necessary, but I don't understand its meaning

**Comment by Njzjhd** [comment]
In addition, I would like to discuss one thing with the author. The above study addresses the classification problem. Although it is a reproduction of the paper, if it is for the regression problem, could the fitness function be modified as follows.
"
def T_fitness(x):
    if x.ndim==1:
        x = x.reshape(1, -1)
    loss = np.zeros(x.shape[0])

    for i in range(x.shape[0]):
        if np.sum(x[i, :]) > 0:
            knn = LGBMRegressor(random_state=44).fit(x_train[:, x[i, :]], y_train)

            loss[i] = r2_score(y_test,knn.predict(X_test[:, x[i, :].astype(bool)]))
        else:
            loss[i] = np.inf
            print(666)
    return loss
"
For the evaluation function in skelern, the input format is
"" acc(y_test, y_pre)",
Please take note

**Comment by ZongSingHuang** [comment]
> For " for i in range(x.shape[0]):", i seems to be equal to 0 only. It seems to be impossible to implement a for loop, the author can do the test. Also what is the meaning of the code " if x.ndim==1: x = x.reshape(1, -1) loss = np.zeros(x.shape[0]) #array([0.])"? This code must be necessary, but I don't understand its meaning

1.
`for i in range(x.shape[0]):`
x is population matrix, row is number of agent, column is feature

2. 
`    if x.ndim==1:
        x = x.reshape(1, -1)`
This is a protective measure if x is a single agnet and dim=1, I don’t remember a bit of a habit I developed when implementing which paper

**Comment by Njzjhd** [comment]
for i in range(x.shape[0]):
For the reproduced code, it seems that there is always only one line in '' range(x.shape[0])'''. That means i is 0 and nothing else. I printed it out and looked at it

**Comment by ZongSingHuang** [comment]
> In addition, I would like to discuss one thing with the author. The above study addresses the classification problem. Although it is a reproduction of the paper, if it is for the regression problem, could the fitness function be modified as follows.
> "
> def T_fitness(x):
> if x.ndim==1:
> x = x.reshape(1, -1)
> loss = np.zeros(x.shape[0])
> 
> ```
> for i in range(x.shape[0]):
>     if np.sum(x[i, :]) > 0:
>         knn = LGBMRegressor(random_state=44).fit(x_train[:, x[i, :]], y_train)
> 
>         loss[i] = r2_score(y_test,knn.predict(X_test[:, x[i, :].astype(bool)]))
>     else:
>         loss[i] = np.inf
>         print(666)
> return loss
> ```
> 
> "
> For the evaluation function in skelern, the input format is
> "" acc(y_test, y_pre)",
> Please take note

Of course, below are some examples, but when it comes to company secrets, I can only show part of the content
for MLR
![image](https://user-images.githubusercontent.com/58209813/105496516-8a7c9f80-5cf8-11eb-9e91-7dc6f32f1949.png)
for SVR(rbf)
![image](https://user-images.githubusercontent.com/58209813/105496720-d4fe1c00-5cf8-11eb-9eef-f429b9e52fc1.png)

reference:
A wrapper approach-based key temperature point selection and thermal error modeling method
A distributed PSO–SVM hybrid system with feature selection and parameter optimization
Particle Swarm Optimization-Based Support Vector Regression for Tourist Arrivals Forecasting
Feature selection and parameter optimization of support vector regression for electric load forecasting

**Comment by ZongSingHuang** [comment]
> for i in range(x.shape[0]):
> For the reproduced code, it seems that there is always only one line in '' range(x.shape[0])'''. That means i is 0 and nothing else. I printed it out and looked at it

Because it involves the author’s original design architecture, I designed the loop in BWOA.py
![image](https://user-images.githubusercontent.com/58209813/105497670-1fcc6380-5cfa-11eb-9983-26b2182aba00.png)

If you are interested, you can look at my other projects, such as
[S-shaped-Binary-Whale-Optimization-Algorithm](https://github.com/ZongSingHuang/S-shaped-Binary-Whale-Optimization-Algorithm)

**Comment by Njzjhd** [comment]
Thank you very much,&nbsp;


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ZongSingHuang/Binary-Whale-Optimization-Algorithm"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年1月22日(星期五) 晚上9:40
收件人:&nbsp;"ZongSingHuang/Binary-Whale-Optimization-Algorithm"<Binary-Whale-Optimization-Algorithm@noreply.github.com&gt;;
抄送:&nbsp;"至爱❤️"<2318109878@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [ZongSingHuang/Binary-Whale-Optimization-Algorithm] Ask a question? (#1)





  
for i in range(x.shape[0]):
 For the reproduced code, it seems that there is always only one line in '' range(x.shape[0])'''. That means i is 0 and nothing else. I printed it out and looked at it
  
Because it involves the author’s original design architecture, I designed the loop in BWOA.py
 
 
If you are interested, you can look at my other projects, such as
 S-shaped-Binary-Whale-Optimization-Algorithm

**Comment by ZongSingHuang** [comment]
> Thank you very much,&nbsp;
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "ZongSingHuang/Binary-Whale-Optimization-Algorithm" <notifications@github.com&gt;; 发送时间:&nbsp;2021年1月22日(星期五) 晚上9:40 收件人:&nbsp;"ZongSingHuang/Binary-Whale-Optimization-Algorithm"<Binary-Whale-Optimization-Algorithm@noreply.github.com&gt;; 抄送:&nbsp;"至爱❤️"<2318109878@qq.com&gt;;"Author"<author@noreply.github.com&gt;; 主题:&nbsp;Re: [ZongSingHuang/Binary-Whale-Optimization-Algorithm] Ask a question? (#1) for i in range(x.shape[0]): For the reproduced code, it seems that there is always only one line in '' range(x.shape[0])'''. That means i is 0 and nothing else. I printed it out and looked at it Because it involves the author’s original design architecture, I designed the loop in BWOA.py If you are interested, you can look at my other projects, such as S-shaped-Binary-Whale-Optimization-Algorithm — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

You welcome
